Add feature flag support to Pages

joshpuetz opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds support to gating access to pages created via `/internal/pages` via feature flags! Adding a feature flag at `/internal/feature_flags` with a name in the format of `page_#{page.slug}` will guard access to that page.

Changes in this PR include:
- Feature flag indicator added to page form
- Flag check in both `PagesController` and `StoriesController`
- Refactored the `FeatureFlag` helper


## Related Tickets & Documents
#7996 - Add Feature Flag support to DEV

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

Feature flag not found matching slug name:

<img width="755" alt="Pages" src="https://user-images.githubusercontent.com/37842/83300156-161b4c00-a1bd-11ea-9a49-a55c26bd0471.png">

Feature flag present matching slug name:
<img width="682" alt="Pages" src="https://user-images.githubusercontent.com/37842/83300169-1b789680-a1bd-11ea-9ce9-e1c06139f757.png">



## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![same page](https://media.giphy.com/media/vO43NDjeQPaQo/giphy.gif)

citizen428 commented on app/controllers/pages_controller.rb:
```diff
@@ -4,6 +4,7 @@ class PagesController < ApplicationController
 
   def show
     @page = Page.find_by!(slug: params[:slug])
+    not_found unless FeatureFlag.enabled?(@page.feature_flag, current_user)
```
This is just an idea, but what if we remove the feature flag name from the form and replace it with a check box that says "Protected by feature flag" or similar? We can then default the flag name to something like `page_#{slug}` which means we could perform this change without having to load a page record from the DB. Potentially we wouldn't even need to add a column to the `pages` table, just a method that returns the name from the slug. 

This would be easier for site admins (only one action instead of two), would save us an extra DB access and they only minor drawback would be that the flag name isn't fully configurable.

joshpuetz replied:
I **love** this idea @citizen428! The only downsides are a) can't customize the flag name like you mentioned, and b) it wouldn't work if you had multiple pages you wanted guarded by a single feature flag.

Since I don't have those requirements for Codeland (which is driving this feature), I'm going to go for it!

citizen428 replied:
Sorry, I meant additional reads in the case the page is disabled. If we could check the flag independent of the page object, this read `@page = Page.find_by!(slug: params[:slug]` becomes unncessary in case the site is blocked by the flag. 

Anyway, you're probably right, like I said, it was just an idea 😃

citizen428 commented on app/helpers/feature_flag.rb:
```diff
@@ -1,7 +1,15 @@
 module FeatureFlag
   extend self # rubocop:disable Style/ModuleFunction
```
I forgot to comment on this in the original PR: we currently do this in several different ways across the code-base:

1. `extend self`
2. Prefixing every method name with `self.`
3. `class << self`
4. `module_function`

For the sake of consistency, we should probably settle on one and given that the one used here needs a temporary Rubocop override I'd vote for 2. or 3. (unless you need the extra feature of `module_function`, in which case 4. is ok).

joshpuetz replied:
Good point: I pretty much took this from the definition of `Flipper::DSL`. I'll opt for option 2 :)

citizen428 commented on app/helpers/feature_flag.rb:
```diff
@@ -1,7 +1,15 @@
+
   def enabled?(feature_name, *args)
-    Flipper[feature_name].enabled?(*args)
+    !exist?(feature_name) || strict_enabled?(feature_name, *args)
```
Maybe we can do something like instead:

```ruby
def enabled?(feature_name, *args)
  Flipper.enabled?(feature_name, *args)
end

def exist?(feature_name)
  Flipper.exist?(feature_name)
end

def active?
  # what the current enabled? is trying to do
end
```

If we do this we can also use delegates to wrap the `Flipper` class methods:

```ruby
module FeatureFlag
  class << self
    delegate :enabled?, :exists?, to: Flipper

    def active?
      ...
    end
  end
end 
```

BTW, not sure about returning true for missing features, that seems a bit counterintuitive to me. The only way I'd expect something to not show up is when it's (1) gated AND (2) enabled.

joshpuetz replied:
How could I have forgotten about `delegate` 🤦 Thanks, this reads much cleaner.

I'm pretty firm that the name of the class should remain FeatureFlag: a Feature is a much more generic term. I like how specific FeatureFlag is.

citizen428 replied:
> I'm pretty firm that the name of the class should remain FeatureFlag

Sure, let's stick to the name `FeatureFlag`. I sometimes just throw ideas out during reviews, unless I use the "request changes" feature (almost never) it's always meant as a suggestion that can be ignored.

citizen428 commented on spec/helpers/feature_flag_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+    context "when flag doesn't exist" do
+      it "returns true" do
+        expect(described_class.enabled?("missing_flag")).to be_truthy # rubocop:disable Rspec/PredicateMatcher
```
If we return a truthy value for pages where we don't have feature flags defined, wouldn't that disable all existing pages unless we retroactively add flags for them?

joshpuetz commented on app/helpers/feature_flag.rb:
```diff
@@ -1,7 +1,15 @@
+
   def enabled?(feature_name, *args)
-    Flipper[feature_name].enabled?(*args)
+    !exist?(feature_name) || strict_enabled?(feature_name, *args)
```
The default behavior of Flipper is to return false is a flag isn't defined. I've added this helper method to return true if:

feature flag is globally enabled
feature flag is disabled, but enabled for the specified user id
feature flag has not been defined
I need input on this last point: if you were to call FeatureFlag.enabled? for a flag that didn't exist...what would you expect to get?

fdoxyz replied:
Oh I see, that check totally went over my head. 

From that perspective I'd maybe say it could be good to make `FeatureFlag.enabled?` return `false` for a flag that doesn't exist.

My reasoning is that if someone sets a check for a feature_flag with a typo then the page will leak as generally available (regardless if @citizen428 idea of using the slug is implemented or not).

joshpuetz replied:
Hmm, that's a fair point. I think I'm covered with this logic for `FeatureFlag.accessible?`

feature flag is globally enabled: **returns true**
feature flag is disabled, but enabled for the specified user id: **returns true**
feature flag specified doesn't exist: **returns false**
nil or empty string passed in: **returns true**

joshpuetz commented:
Oof, good catch @fdoxyz! (https://github.com/thepracticaldev/dev.to/pull/8149#pullrequestreview-421041265)

I think I'll move the null check inside `FeatureFlag.accessible?`: I hate how having null checks pepper throughout the code looks!

mstruve commented on app/controllers/stories_controller.rb:
```diff
@@ -111,8 +111,12 @@ def handle_user_or_organization_or_podcast_or_page_index
-      Honeycomb.add_field("stories_route", "page")
-      handle_page_display
+      if FeatureFlag.enabled?(@page.feature_flag, current_user)
+        Honeycomb.add_field("stories_route", "page")
```
🙏  Thanks for adding this

mstruve commented on db/schema.rb:
```diff
@@ -794,6 +794,7 @@
     t.text "body_markdown"
     t.datetime "created_at", null: false
     t.string "description"
+    t.string "feature_flag"
```
Rather than add a column here, it would be nice to keep the feature flags contained to just Flipper, why not use the slug as the flag instead? The slug will ensure uniqueness and is very unlikely to change for the life of the page.

joshpuetz replied:
Feature flagging a page is going to be optional: we don't want to ever _require_ a feature flag: they are meant to be short lived. I'm going to use the slug name as the flag name as @citizen428 suggested  (https://github.com/thepracticaldev/dev.to/pull/8149#discussion_r432276233), but will still need to record the fact this page is guarded SOMEWHERE.

joshpuetz commented:
@mstruve @citizen428 I hear your concerns about flag data leaking into the database. I've updated the PR to a) not store _anything_ in the database, and b) alert the admin if a feature flag matching `page_#{slug_name}` is present. I've added text and links at the bottom of the page forma to alert the user if the flag is present or not!

citizen428 commented on spec/helpers/feature_flag_spec.rb:
```diff
@@ -0,0 +1,66 @@
+    let(:user) { UserStruct.new(flipper_id: 1) }
+
+    it "returns false when flag doesn't exist" do
+      expect(described_class.accessible?("missing_flag")).to be_truthy # rubocop:disable Rspec/PredicateMatcher
```
No blocker for this PR, but for a future refactor it may be easier to move the `rubocop:disable` to the top and add a `rubocop:enable Rspec/PredicateMatcher` at the bottom of the file as we do in some other files.

https://rubocop.readthedocs.io/en/latest/configuration/#disabling-cops-within-source-code

citizen428 commented on app/helpers/feature_flag.rb:
```diff
@@ -1,7 +1,9 @@
-  def enabled?(feature_name, *args)
-    Flipper[feature_name].enabled?(*args)
+    def accessible?(feature_flag_name, *args)
+      feature_flag_name.blank? || !exist?(feature_flag_name) || enabled?(feature_flag_name, *args)
```
Right now we call `accessible?` in `StoriesController`, where we pass in `@page.feature_flag_name`. In `Page` we define this method to return `"page_#{slug}"`. So am I missing something or can this never be `blank?`?

joshpuetz replied:
You're right: the way I'm calling it now it won't be blank, but I want this helper class to be resilient for any kind of use case: I don't want to always assume feature flag name will be passed in!
